keps/sig-node/2837-pod-level-resource-spec/README.md

```diff
@@ -0,0 +1,1108 @@
+
+
+### Non Goals
+
+
+1. This KEP focuses on the core resource types of CPU and memory. It doesn’t intend to address other resource types (e.g. GPU, storage, network bandwidth)
+at pod level in this phase. However, it could be considered in future extensions of the KEP.
+
+
+## Proposal
+
+
+### User Stories
+
+
+#### Story 1
+
+A development environment implemented as a Kubernetes Pod allows for separation of tools and a (web-based) IDE between multiple side-cars.
+The development environment defines a container with the web-server serving the IDE itself, and constrains it to use a certain amount of memory.
+Additional tools are provided in additional side-car deployments - for example an LSP( https://langserver.org/) service, a terminal, and more. 
+Using this new feature, the containers providing the terminal, the LSP services, and the set of tools being utilized can share the resource limit
+defined for the pod. Consider the following pod definition:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  labels:
+    run: ide
+  name: myide
+spec:
+  resources:
+    limits:
+      memory: "1024Mi"
+      cpu: "4"
+  containers:
+  - image: shell
+    name: debian:buster
+  - image: tool1
+    name: first-tool:latest
+  - image: tool2
+    name: second-tool:latest
+  - name: ide
+    image: theia:latest
+    resources:
+      requests:
+        memory: 128M
+        cpu: "0.5"
+      limits:
+        memory: 256M
+        cpu: "1"
+```
+
+Using the pod-level resource definition enables the `tool1` and `tool2` containers to be constrained by the total limit for the pod. 
+
+Without this feature, the developer would need to decide a-priori how much resources to allocate to the tools - and this is not easy to do for this workload.
+
+
+#### Story 2
+
+Consider a scenario where a pod has some containers with occasional resource spikes while other containers often have idle resources. With current container 
+level resource settings, the peak usage of each container needs to be accounted for when setting limits. This leads to wasted resources when the spikes are 
+not occurring. Pod-level resource limits will allow setting a limit for the entire pod, enabling containers to share resources more efficiently without the
+need to over-provision for each container’s peak usage. 
+
+
+### Notes/Constraints/Caveats
+
+
+1. Community is trying to move cgroupv1 to maintenance mode in future Kubernetes versions (likely 1.31). Hence this feature will be supported only on cgroupv2.
+2. This feature is an **opt-in** feature, and will have no effect on existing deployments. Only deployments that explicitly require this functionality should
+turn it on by specifying the relevant resource section at pod-level in the Pod specification.
+
+
+### Risks and Mitigations
+
+Since this is an **opt-in** feature, there should be no risk in merging the changes. Existing workloads that do not use this feature won’t be affected.
```

Comment:
> Existing workloads that do not use this feature won’t be affected.

Theoretically they can if we will apply bad limits to the Pods which opted-in for the feature and those bad limits create a noisy neighbour situation for existing workloads

Reply:
this is definitely something to cover in rollout section of PRR

Reply:
I added a sentence to address this. I feel the problem of bad limits to new pods creating a noisy neighbour is not limited to this feature. It could also happen is bad container level limits are applied.